bigframes/session/__init__.py

```diff
@@ -1297,34 +1167,51 @@ def _create_empty_temp_table(
-    def _create_sequential_ordering(
+    def _create_total_ordering(
         self,
         table: ibis_types.Table,
-        index_cols: Iterable[str] = (),
-        api_name: str = "",
-    ) -> Tuple[ibis_types.Table, orderings.ExpressionOrdering]:
+    ) -> core.ArrayValue:
         # Since this might also be used as the index, don't use the default
         # "ordering ID" name.
-        default_ordering_name = guid.generate_guid("bigframes_ordering_")
-        default_ordering_col = (
-            ibis.row_number().cast(ibis_dtypes.int64).name(default_ordering_name)
+        ordering_hash_part = guid.generate_guid("bigframes_ordering_")
+        ordering_rand_part = guid.generate_guid("bigframes_ordering_")
+
+        str_values = list(
+            map(lambda col: _convert_to_string(table[col]), table.columns)
         )
-        table = table.mutate(**{default_ordering_name: default_ordering_col})
-        table_ref = self._ibis_to_session_table(
-            table,
-            cluster_cols=list(index_cols) + [default_ordering_name],
-            api_name=api_name,
+        full_row_str = (
+            str_values[0].concat(*str_values[1:])
+            if len(str_values) > 1
+            else str_values[0]
         )
-        table = self.ibis_client.table(
-            f"{table_ref.project}.{table_ref.dataset_id}.{table_ref.table_id}"
+        full_row_hash = full_row_str.hash().name(ordering_hash_part)
+        # Used to disambiguate between identical rows (which will have identical hash)
+        random_value = ibis.random().name(ordering_rand_part)
+
+        # Might need to actually reselect here?
+        original_column_ids = table.columns
+        table_with_ordering = table.select(
+            itertools.chain(original_column_ids, [full_row_hash, random_value])
         )
-        ordering_reference = core.OrderingColumnReference(default_ordering_name)
+
+        ordering_ref1 = core.OrderingColumnReference(ordering_hash_part)
+        ordering_ref2 = core.OrderingColumnReference(ordering_rand_part)
         ordering = orderings.ExpressionOrdering(
-            ordering_value_columns=tuple([ordering_reference]),
-            total_ordering_columns=frozenset([default_ordering_name]),
-            integer_encoding=IntegerEncoding(is_encoded=True, is_sequential=True),
+            ordering_value_columns=(ordering_ref1, ordering_ref2),
+            total_ordering_columns=frozenset([ordering_hash_part, ordering_rand_part]),
+        )
+        columns = [table_with_ordering[col] for col in original_column_ids]
+        hidden_columns = [
+            table_with_ordering[ordering_hash_part],
+            table_with_ordering[ordering_rand_part],
+        ]
+        return core.ArrayValue.from_ibis(
+            self,
+            table_with_ordering,
+            columns,
+            hidden_ordering_columns=hidden_columns,
+            ordering=ordering,
         )
-        return table, ordering
 
     def _ibis_to_session_table(
```

Comment:
We can remove this method now, I believe.

Reply:
Still used by ArrayValue.cached(). Renaming to the more accurate `_ibis_to_temp_table` since the session dataset isn't being used anymore.